File: Semester 4/Class13_Computer_Contest_Level_3_Notes_and_Homework_Dec_03/lvl3_hw_sol/cipele.java
import java.util.*;
import java.io.*;
public class homework {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int n, m, lft[], rit[];
    public static void main(String[] args) throws IOException {
    	n = readInt(); m = readInt(); lft = new int[n]; rit = new int[m];
    	for(int i=0; i<n; i++) lft[i] = readInt();
    	for(int i=0; i<m; i++) rit[i] = readInt();
    	Arrays.sort(lft); Arrays.sort(rit);
    	int lo = 0, hi = (int)1e9, ans = 0;
    	while(lo <= hi) {
    		int mid = (lo + hi)/2;
    		if(check(mid)) { ans = mid; hi = mid - 1; }
    		else lo = mid + 1;
    	}
    	System.out.println(ans);
    }
    static boolean check(int limit) {
    	int pair = 0;
    	for(int i=0, j=0; i<n && j<m; ) {
    		if(Math.abs(lft[i] - rit[j]) <= limit) {
    			i++; j++; pair++;
    		}else if(lft[i] < rit[j]) {
    			i++;
    		}else j++;
    	}
    	return pair == Math.min(m, n);
    }
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: Semester 4/Class13_Computer_Contest_Level_3_Notes_and_Homework_Dec_03/lvl3_hw_sol/cipele.cpp
#include <bits/stdc++.h>
using namespace std;
const int MM = 1e5+5;
int n, m, ans; vector<int> lft, rit;
bool check(int lmt){
    int cnt = 0;
    for(int i=0, j=0; i<lft.size() && j < rit.size(); ){
        if(abs(lft[i] - rit[j]) <= lmt) {
            i++; j++; cnt++;
        }else if(lft[i] < rit[j]) {
            i++;
        }else {
            j++;
        }
    }
    return cnt == min(n, m);
}
int main(){
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> n >> m;
    for(int i=0, x; i<n; i++) {
        cin >> x; lft.push_back(x);
    }
    for(int i=0, x; i<m; i++) {
        cin >> x; rit.push_back(x);
    }
    sort(lft.begin(), lft.end());
    sort(rit.begin(), rit.end());
    int lo = 0, hi = 1e9;
    while(lo <= hi){
        int mid = (lo + hi)/2;
        if(check(mid)) { ans = mid; hi = mid - 1;}
        else lo = mid + 1;
    }
    cout << ans << "\n";
}